# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/PersistentBufferProvider.h
# Commit: 31935f164ef0
# Full Hash: 31935f164ef0d6c4668f211482c139847dcab7f2
# Author: Lee Salzman <lsalzman@mozilla.com>
# Date: 2023-12-13 00:43:26
# Regressor Bug: 1829026
# File Overlap Count: 1
# Description:
#   Bug 1829026 - Update CanvasTranslator to work with DrawTargetWebgl. r=aosmond
#   
#   This adds the necessary infrastructure for CanvasTranslator to allocate DrawTargetWebgl
#   instead of just allocating TextureData, and to use RemoteTextureMap to handle sending
#   the DrawTargetWebgl frames to the compositor.
# ==============================================================================

diff -r 13c9bf69085f -r 31935f164ef0 gfx/layers/PersistentBufferProvider.h
--- a/gfx/layers/PersistentBufferProvider.h	Tue Dec 12 07:35:02 2023 +0000
+++ b/gfx/layers/PersistentBufferProvider.h	Tue Dec 12 07:35:03 2023 +0000
@@ -24,11 +24,11 @@
 namespace gfx {
 class SourceSurface;
 class DrawTarget;
-class DrawTargetWebgl;
 }  // namespace gfx
 
 namespace layers {
 
+struct RemoteTextureOwnerId;
 class TextureClient;
 
 /**
@@ -104,12 +104,7 @@
    */
   virtual bool RequiresRefresh() const { return false; }
 
-  /**
-   * Provide a WebGL front buffer for compositing, if available.
-   */
-  virtual Maybe<layers::SurfaceDescriptor> GetFrontBuffer() {
-    return Nothing();
-  }
+  virtual Maybe<SurfaceDescriptor> GetFrontBuffer() { return Nothing(); }
 };
 
 class PersistentBufferProviderBasic : public PersistentBufferProvider {
@@ -146,18 +141,17 @@
   RefPtr<gfx::SourceSurface> mSnapshot;
 };
 
-class PersistentBufferProviderAccelerated
-    : public PersistentBufferProviderBasic {
+class PersistentBufferProviderAccelerated : public PersistentBufferProvider {
  public:
   MOZ_DECLARE_REFCOUNTED_VIRTUAL_TYPENAME(PersistentBufferProviderAccelerated,
                                           override)
 
-  explicit PersistentBufferProviderAccelerated(gfx::DrawTarget* aTarget);
+  static already_AddRefed<PersistentBufferProviderAccelerated> Create(
+      gfx::IntSize aSize, gfx::SurfaceFormat aFormat,
+      KnowsCompositor* aKnowsCompositor);
 
   bool IsAccelerated() const override { return true; }
 
-  Maybe<layers::SurfaceDescriptor> GetFrontBuffer() override;
-
   already_AddRefed<gfx::DrawTarget> BorrowDrawTarget(
       const gfx::IntRect& aPersistedRect) override;
 
@@ -166,14 +160,27 @@
   already_AddRefed<gfx::SourceSurface> BorrowSnapshot(
       gfx::DrawTarget* aTarget) override;
 
+  void ReturnSnapshot(already_AddRefed<gfx::SourceSurface> aSnapshot) override;
+
+  bool PreservesDrawingState() const override { return true; }
+
+  void OnShutdown() override { Destroy(); }
+
+  Maybe<SurfaceDescriptor> GetFrontBuffer() override;
+
   bool RequiresRefresh() const override;
 
-  void OnMemoryPressure() override;
-
  protected:
+  explicit PersistentBufferProviderAccelerated(
+      const RefPtr<TextureClient>& aTexture);
   ~PersistentBufferProviderAccelerated() override;
 
-  gfx::DrawTargetWebgl* GetDrawTargetWebgl() const;
+  void Destroy();
+
+  RefPtr<TextureClient> mTexture;
+
+  RefPtr<gfx::DrawTarget> mDrawTarget;
+  RefPtr<gfx::SourceSurface> mSnapshot;
 };
 
 /**